wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt

```diff
@@ -0,0 +1,487 @@
+package com.squareup.wire.kotlin
+
+import android.os.Parcel
+import android.os.Parcelable
+import com.squareup.kotlinpoet.ARRAY
+import com.squareup.kotlinpoet.AnnotationSpec
+import com.squareup.kotlinpoet.BOOLEAN
+import com.squareup.kotlinpoet.ClassName
+import com.squareup.kotlinpoet.CodeBlock
+import com.squareup.kotlinpoet.DOUBLE
+import com.squareup.kotlinpoet.FLOAT
+import com.squareup.kotlinpoet.FunSpec
+import com.squareup.kotlinpoet.INT
+import com.squareup.kotlinpoet.KModifier.DATA
+import com.squareup.kotlinpoet.KModifier.OVERRIDE
+import com.squareup.kotlinpoet.KModifier.PRIVATE
+import com.squareup.kotlinpoet.LONG
+import com.squareup.kotlinpoet.NameAllocator
+import com.squareup.kotlinpoet.ParameterSpec
+import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
+import com.squareup.kotlinpoet.PropertySpec
+import com.squareup.kotlinpoet.TypeName
+import com.squareup.kotlinpoet.TypeSpec
+import com.squareup.kotlinpoet.asClassName
+import com.squareup.wire.EnumAdapter
+import com.squareup.wire.FieldEncoding
+import com.squareup.wire.ProtoAdapter
+import com.squareup.wire.ProtoReader
+import com.squareup.wire.ProtoWriter
+import com.squareup.wire.WireEnum
+import com.squareup.wire.schema.EnumType
+import com.squareup.wire.schema.Field
+import com.squareup.wire.schema.MessageType
+import com.squareup.wire.schema.ProtoFile
+import com.squareup.wire.schema.ProtoType
+import com.squareup.wire.schema.Schema
+import com.squareup.wire.schema.Type
+import okio.ByteString
+import java.util.*
+
+class KotlinGenerator private constructor(
+    val schema: Schema,
+    private val nameToKotlinName: Map<ProtoType, ClassName>,
+    private val emitAndroid: Boolean
+) {
+  /** Returns the Kotlin type for [protoType]. */
+  fun typeName(protoType: ProtoType) = nameToKotlinName.getValue(protoType)
+
+  /** Returns the full name of the class generated for `type`.  */
+  fun generatedTypeName(type: Type) = typeName(type.type())
+
+  fun generateType(type: Type): TypeSpec = when (type) {
+    is MessageType -> generateMessage(type)
+    is EnumType -> generateEnum(type)
+    else -> error("Unknown type $type")
+  }
+
+  // TODO Make this a CACHE
```

Comment:
Cache is probably the wrong way to think about this. The LoadingCache that we use in JavaGenerator is more like a computing map than a cache; it never evicts and we rely on it for correctness!